core/jvm/src/main/scala/scalaz/zio/RTS.scala

```diff
@@ -204,9 +204,9 @@ private object RTS {
           result.set(v)
 
         case _ =>
-          while (result.get eq null) {
-            result.synchronized {
-              if (result.get eq null) result.wait()
+          result.synchronized {
```

Comment:
The reason for the old code is that you can avoid the overhead of `synchronized` in the case where the `IO` evaluated synchronously. It's an old trick to cut down on the need for locking (in a smaller percentage of cases).

Reply:
Oooh optimisation had crossed my mind, thanks for the explanation 👍